webrender/src/renderer.rs

```diff
@@ -514,107 +512,79 @@ impl CacheTexture {
+        let pbo = device.create_pbo();
 
-        GpuDataTexture {
+        VertexDataTexture {
             id,
-            layout: PhantomData,
+            pbo,
         }
     }
 
-    fn init<T: Default>(&mut self,
-                        device: &mut Device,
-                        data: &mut Vec<T>) {
+    fn update<T>(&mut self,
+                 device: &mut Device,
+                 data: &mut Vec<T>) {
         if data.is_empty() {
             return;
         }
 
-        let items_per_row = L::items_per_row::<T>();
-        let rows_per_item = L::rows_per_item::<T>();
+        debug_assert!(mem::size_of::<T>() % 16 == 0);
+        let texels_per_item = mem::size_of::<T>() / 16;
+        let items_per_row = MAX_VERTEX_TEXTURE_WIDTH / texels_per_item;
 
         // Extend the data array to be a multiple of the row size.
         // This ensures memory safety when the array is passed to
         // OpenGL to upload to the GPU.
         if items_per_row != 0 {
             while data.len() % items_per_row != 0 {
-                data.push(T::default());
+                data.push(unsafe { mem::uninitialized() });
             }
         }
 
-        let height = if items_per_row != 0 {
-            data.len() / items_per_row
-        } else {
-            data.len() * rows_per_item
-        };
+        let width = (MAX_VERTEX_TEXTURE_WIDTH - (MAX_VERTEX_TEXTURE_WIDTH % texels_per_item)) as u32;
+        let needed_height = (data.len() / items_per_row) as u32;
 
-        device.init_texture(self.id,
-                            L::texture_width::<T>() as u32,
-                            height as u32,
-                            L::image_format(),
-                            L::texture_filter(),
-                            RenderTargetMode::None,
-                            1,
-                            Some(unsafe { mem::transmute(data.as_slice()) } ));
-    }
-}
+        // Determine if the texture needs to be resized.
+        let texture_size = device.get_texture_dimensions(self.id);
 
-pub struct VertexDataTextureLayout {}
+        if needed_height > texture_size.height {
+            let new_height = (needed_height + 127) & !127;
 
-impl GpuStoreLayout for VertexDataTextureLayout {
-    fn image_format() -> ImageFormat {
-        ImageFormat::RGBAF32
-    }
+            device.init_texture(self.id,
+                                width,
+                                new_height,
+                                ImageFormat::RGBAF32,
+                                TextureFilter::Nearest,
+                                RenderTargetMode::None,
+                                1,
+                                None);
+        }
 
-    fn texture_width<T>() -> usize {
-        MAX_VERTEX_TEXTURE_WIDTH - (MAX_VERTEX_TEXTURE_WIDTH % Self::texels_per_item::<T>())
-    }
+        // Bind a PBO to do the texture upload.
+        // Updating the texture via PBO avoids CPU-side driver stalls.
```

Comment:
I don't think the comment really applies to this use case: if your `update_texture_from_pbo` follows `update_pbo_data`, then the actual work the driver has got to do is no different from a regular `glTexImage*`, and it will still stall.
In order to get the benefit from PBOs, we need to:
  1. do the `update_pbo_data` as early as possible, giving it more time until the PBO -> texture copy is issued
  2. don't re-use the same PBO space until the *GPU* is done with the previous frame, which means having a number of those PBO buffers cycled over

Reply:
Intuitively, I agree with your comment above - but the ```perf``` output from the intel driver suggests otherwise.

With the path above, I see no performance stalls reported, and nothing in the profile graphs. With the previous code, we see stalls in the perf logs, and noticeable stalls in the frame profiler.

I guess the driver has a stall-free path for PBOs but not for textures, perhaps?

Reply:
Thanks for confirming this! I'll try to get a close look what's going on there when I get a chance.